pkg/controller/servicebinding/secret.go

```diff
@@ -27,17 +34,46 @@ func (s *secret) buildResourceClient() dynamic.ResourceInterface {
 	return s.client.Resource(gvr).Namespace(s.ns)
 }
 
-// createOrUpdate will take informed payload and either create a new secret or update an existing
-// one. It can return error when Kubernetes client does.
-func (s *secret) createOrUpdate(payload map[string][]byte, ownerReference metav1.OwnerReference) (*unstructured.Unstructured, error) {
+// HashedResourceNameForData is a utility function to get a checksum of the resource data and append it to resource name
+func HashedResourceNameForData(name string, data map[string][]byte) (string, error) {
+	keys := make([]string, 0, len(data))
+	for k := range data {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	hash := sha1.New()
+	for _, k := range keys {
+		_, err := hash.Write([]byte(k))
+		if err != nil {
+			return "", err
+		}
+		_, err = hash.Write(data[k])
+		if err != nil {
+			return "", err
+		}
+	}
+	dataHash := hex.EncodeToString(hash.Sum(nil))
+
+	return name + "-" + dataHash, nil
+}
+
+// create will take informed payload and create a new secret
+// It can return error when Kubernetes client does.
+func (s *secret) createOrUpdate() (*unstructured.Unstructured, error) {
```

Comment:
@pedjak Please refer to the conversations above by @isutton regarding the logic of why the comparison is done later in the function

Reply:
> with the applied naming strategy, secret are immutable, i.e. they are never updated. Hence, this function could be called `maybeCreate` to reflect this. Given the immutability, the logic of the function can be simplified:

This is not accurate. If an external actor modifies the secret, the secret is never fixed since there's no `Update` call to fix it.

It's immutable from the *Operator* point of view, but it is mutable from the * API server* point of view.

The contract indicates the secret is owned by the controller, so any external interferences should be corrected, and the easiest heuristics is to overwrite the secret with the correct content.

Reply:
> This is not accurate. If an external actor modifies the secret, the secret is never fixed since there's no Update call to fix it.

Even if we fix the secret in the cluster, we will not be able to propagate the secret changes to the application, because the secret reference in the application resource will not change.

IMHO, guarding against such case is something we should not worry here. We should think about addressing it only if this becomes relevant for users.

Reply:
> > This is not accurate. If an external actor modifies the secret, the secret is never fixed since there's no Update call to fix it.
> 
> Even if we fix the secret in the cluster, we will not be able to propagate the secret changes to the application, because the secret reference in the application resource will not change.

Those secrets should be considered Service Binding Operator internal data and, if this data is corrupt by any reason, the Operator should be able to fix it. I can't help but stress that we're building on top of a self-healing system and Operators we write should attempt to self-heal as much as possible.

Getting to whether the workload picks those changes up, this is an orthogonal issue that a) might be already solved in the case of Deployments or b) another actor could be created to solve this specifically for the use cases we're actively showcasing.

> 
> IMHO, guarding against such case is something we should not worry here. We should think about addressing it only if this becomes relevant for users.

I disagree; we should address reliability and self-healing issues when first learn about them.

Reply:
> I disagree; we should address reliability and self-healing issues when first learn about them.

Fine, my idea was to solve it later, but if you insist we can address it now.